crates/types/src/string.rs

```diff
@@ -201,16 +180,134 @@ impl StringBuilder {
+        self.reserve_raw(new_capacity);
+    }
+
+    /// Reserve space for exactly N more bytes.
+    ///
+    /// Does not allocate if enough space is available.
+    pub fn reserve_exact(&mut self, aditional: usize) {
+        if let Some(new_capacity) = self.min_capacity_for_additional(aditional)
+        {
+            self.reserve_raw(new_capacity);
+        }
+    }
+
+    /// Allocate new_capacity bytes.
+    ///
+    /// Accepts [`NonZeroUsize`] allocation size should never be zero for portability reasons.
+    /// A check isn't included as all call sites should already handle the case.
+    fn reserve_raw(&mut self, new_capacity: NonZeroUsize) {
+        let ptr = unsafe {
+            libc::realloc(
+                self.inner.data as *mut ffi::c_void,
+                new_capacity.get(),
+            )
+        };
+        // realloc may return null if it is unable to allocate the requested memory
+        if ptr.is_null() {
+            unable_to_alloc_memory();
+        }
+        self.inner.data = ptr as *mut ffi::c_char;
+        self.cap = new_capacity.get();
+    }
+
+    /// Finish building the [`String`] but do not shrink the allocation
+    ///
+    /// Useful if we already know that the allocation does not have extra capacity.
+    ///
+    /// In debug mode checks if any of the following are true:
+    /// - data ptr is null with a capacity and length of zero.
+    /// - ptr points to an allocation and capacity > length.
+    /// - ptr points to an allocation with a capacity > 1 (if the capacity is 1 we are creating the
+    ///     allocation just to store a null byte, we should return null instead)
+    #[inline]
+    fn finish(self) -> String {
+        let s = String { data: self.inner.data, len: self.inner.len() };
+
+        // extra sanity check
+        if s.data.is_null() {
+            debug_assert!(s.is_empty());
+            debug_assert_eq!(self.cap, 0);
+        } else {
+            debug_assert!(self.cap > self.inner.len());
+        }
 
         // Prevent self's destructor from being called.
         std::mem::forget(self);
-
         s
     }
+
+    /// Returns the remaining *usable* capacity, i.e. the remaining capacity minus
+    /// the space reserved for the null terminator.
+    #[inline(always)]
+    fn remaining_capacity(&self) -> usize {
+        if self.inner.data.is_null() {
+            debug_assert_eq!(self.inner.len(), 0);
+            return 0;
+        }
+        debug_assert!(
+            self.cap > 0,
+            "when data ptr is not null capacity must always be larger than 0"
+        );
+        debug_assert!(
+            self.cap > self.inner.len(),
+            "allocated capacity must always be bigger than length"
+        );
+        self.cap - self.inner.len() - 1
+    }
+
+    #[inline]
+    fn min_capacity_for_additional(
+        &self,
+        aditional: usize,
+    ) -> Option<NonZeroUsize> {
+        if aditional == 0 {
+            return None;
+        }
```

Comment:
Is there a reason for special-casing this? It's already handled by the following check.